### PR TITLE
Add dict-based indexing to Variable, DataArray, and Dataset (#3711)

### LINKED ISSUE #3711: Set/Get a value by a dictionary of dimensions names and values
If I understood the docs correctly, and in particular (from [here](https://scipp.github.io/user-guide/slicing.html#Limitations)):

> ### Limitations
>
> Label-based indexing not supported for:
>
> - Multi-dimensional coordinates.

The common `.loc` method of `xarray` is missing for me. I managed to monkey-patch `scipp.DataArray` to include equivalent methods:

```python
def sc_get_with_dict(self, index_dict):
    """Get item using dictionary of {dim_name: dim_index}"""
    result = self
    for dim_name, value in index_dict.items():
        result = result[dim_name, value]
    return result
sc.DataArray.get = sc_get_with_dict

def sc_set_with_dict(self, index_dict, val):
    """Set item using dictionary of {dim_name: dim_index}"""
    # Convert index_dict to list of items to process all but the last
    items = list(index_dict.items())
    target = self
    for dim_name, idx_value in items[:-1]:
        target = target[dim_name, idx_value]
    target[*items[-1]] = val
sc.DataArray.set = sc_set_with_dict
```

I was wondering whether anyone has ever felt these methods missing, and whether maybe by design you didn't include them.

### PR DESCRIPTION
**Pull request by doronbehar**
Fixes #3711.

Co-Authored-By: Claude Sonnet 4.6 <noreply@anthropic.com>

### CODE REVIEW

**Comment by doronbehar** [comment]
OK so CI is red not because of me - `dict-getsetitem` must be a valid branch name!

File: tests/dict_indexing_test.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        ('s',),
+        id='s',
+    ),
+    # m, l, s indexed → (N, K) remain
+    pytest.param(
+        {'m': 1, 'l': 3, 's': 2},
+        ('n', 'k'),
+        id='nk',
+    ),
+]
```
Comment (jl-wynen) [inline tests/dict_indexing_test.py]: Can you add a test with an empty dict? I would expect this to satisfy:
```python
result = obj[{}]
assert sc.identical(result, obj)
assert result is not obj
```

Comment (jl-wynen) [inline tests/dict_indexing_test.py]: Can you also add a test for error handling? 

- Bad dim label
- Bad index

File: lib/python/bind_slice_methods.h [hunk 2]
```diff
@@ -350,4 +350,33 @@ void bind_slice_methods(pybind11::class_<T, Ignored...> &c) {
...
+    items.reserve(index.size());
+    for (const auto &item : index) {
+      items.emplace_back(py::reinterpret_borrow<py::object>(item.first),
+                         py::reinterpret_borrow<py::object>(item.second));
+    }
+    py::object target = py::cast(self);
+    for (std::size_t i = 0; i + 1 < items.size(); ++i) {
+      target = target.attr("__getitem__")(
+          py::make_tuple(items[i].first, items[i].second));
+    }
```
Comment (jl-wynen) [inline lib/python/bind_slice_methods.h]: This loop is on both new functions. Can you extract it?

File: lib/python/bind_slice_methods.h [hunk 3]
```diff
@@ -350,4 +350,33 @@ void bind_slice_methods(pybind11::class_<T, Ignored...> &c) {
         return slice_by_list<T>(self, to_dim_type(std::move(indices)));
       },
       py::call_guard<py::gil_scoped_release>());
+  // Dict-based indexing: obj[{'x': 0, 'y': 1:3}] is equivalent to
+  // obj['x', 0]['y', 1:3], applying each (dim, index) pair in insertion order.
+  c.def("__getitem__", [](T &self, const py::dict &index) -> py::object {
+    py::object result = py::cast(self);
+    for (const auto &item : index) {
+      result = result.attr("__getitem__")(
+          py::make_tuple(item.first, item.second));
```
Comment (jl-wynen) [inline lib/python/bind_slice_methods.h]: I'm not too happy with casting to Python types here. Is there a way to do this in C++ directly?

**Review by jl-wynen — COMMENTED** [review]
Thank you for this contribution!

**Comment by jl-wynen** [comment]
> OK so CI is red not because of me - `dict-getsetitem` must be a valid branch name!

Does this happen because you opened this PR from a fork? I guess we need to update our CI config.

**Comment by doronbehar** [comment]
> > OK so CI is red not because of me - `dict-getsetitem` must be a valid branch name!
> 
> Does this happen because you opened this PR from a fork? I guess we need to update our CI config.

I opened the PR from a fork, because I'm not a member of the `scipp` organization :).

File: lib/python/bind_slice_methods.h (hunk 3 above)
Comment (doronbehar) [reply lib/python/bind_slice_methods.h]: I'm not sure I understand this. TBH I don't understand much about mixing C++ with Python etc. I'm just a scientist who likes Python and likes programming, and I'm paying the tokens for the AI :). Anyway, I gave it your comments and it solved them but I'm not able to vouch for whether this particular comment of yours is resolved properly.

File: lib/python/bind_slice_methods.h (hunk 2 above)
Comment (doronbehar) [reply lib/python/bind_slice_methods.h]: Looks like the AI resolved this comment and the latter comment about Python casting in 1 shot.

File: tests/dict_indexing_test.py (hunk 1 above)
Comment (doronbehar) [reply tests/dict_indexing_test.py]: Good idea!

Comment (doronbehar) [reply tests/dict_indexing_test.py]: Sure, done!

Comment (doronbehar) [reply tests/dict_indexing_test.py]: For errors to be checked properly, see:

- https://github.com/scipp/scipp/pull/3926

**Comment by doronbehar** [comment]
All issues above were fixed, but converted to a draft since #3926 is a prerequisite for this PR (this PR's tests to be particular).

**Comment by doronbehar** [comment]
PR is rebased and ready for review.

**Comment by SimonHeybrock** [comment]
Thanks, starting to look into this now, I may push some changes at some point.

**Comment by SimonHeybrock** [comment]
Thanks for getting this rolling @doronbehar — the feature is a good fit, and the test coverage you added was a useful base. I've pushed a commit on top of your branch that reworks the internals; since some of it is subtle C++/pybind territory, it seemed more efficient to push the changes than to spell them out in review comments. Summary of what changed and why:

**A silent data-loss bug in `__setitem__`.** Integer-array indexing (`obj[{'x': [0, 2], ...}]`) returns a *copy* in scipp, not a view. The previous implementation chained `__getitem__` for all but the last dict entry, so an integer array anywhere before the last position made the assignment write into a temporary and silently discard the values. Dict-based `__setitem__` now rejects integer-array indices with a `TypeError` (the tuple-based API never supported them in `__setitem__` either).

**Parse first, then slice.** The dict is now parsed into C++ slice parameters up front, and the actual slicing runs with the GIL released, consistent with the other indexing overloads. Because dict keys are unique, each dimension is indexed at most once, so the result is independent of dict order — the docs now say so instead of promising "insertion order". This also lets us apply view-producing slices before the copying integer-array extractions, keeping copies as small as possible. (This addresses @jl-wynen's earlier comment about dispatching on Python types: the per-item `py::isinstance`/`py::cast` chains are gone from the hot path, and the label-based `__setitem__` special-casing collapsed into the existing `slicer::set` machinery.)

**Edge cases aligned with tuple-based indexing.** numpy integer scalars now work as index values (floats are still rejected), non-string keys and label-indexing a plain `Variable` produce clear errors, and `obj[{}]` returns a full-slice view like `obj[...]` rather than a plain shallow copy.

**`DataGroup` and typing.** Widening `ScippIndex` with `dict` made `DataGroup.__getitem__` claim dict support it didn't have, so it now has it. The stubs were narrowed to what `__setitem__` actually accepts.

I also flattened the test file (the label-based test class needed a fair bit of Python scoping gymnastics — plain parametrization does the same with less machinery) and extended coverage for the cases above.

**Comment by SimonHeybrock** [comment]
@jl-wynen Could you have a look at my changes? Thanks!

**Comment by doronbehar** [comment]
@jl-wynen  :pray:

File: lib/python/bind_slice_methods.h [hunk 4]
```diff
@@ -242,6 +244,62 @@ T slice_by_list(const T &obj,
...
+
+// All slice parameters are computed against `self`. This is valid because
+// each dimension is indexed at most once: slicing one dimension changes
+// neither the extent of the others nor the (1-D) coords used for label-based
+// lookup along them.
+template <class T>
+ParsedDictIndex parse_dict_index(T &self, const py::dict &index) {
+  ParsedDictIndex parsed;
+  for (const auto &item : index) {
+    if (!py::isinstance<py::str>(item.first))
```
Comment (jl-wynen) [inline lib/python/bind_slice_methods.h]: This is more restrictive than indexing with tuples. E.g., this works:
```python
x[np.str_('abc'), 0]
```
but with the proposed code, the following would fail:
```python
x[{np.str_('abc'): 0}]
```

To fix this, you can use a function like this:
```cpp
Dim parse_dict_key_as_dim(const py::handle &key) {
  try {
    return Dim{py::cast<std::string>(key)};
  } catch (const py::cast_error& e) {
    throw except::TypeError(
      "Dict-based indexing requires string keys (dimension labels).");
  }
}
```

As an added bonus, this gets rid of the duplicate type check.

File: lib/python/bind_slice_methods.h [hunk 5]
```diff
@@ -242,6 +244,62 @@ T slice_by_list(const T &obj,
...
+      parsed.slices.push_back(
+          get_slice(self, {dim, py::cast<scipp::index>(val)}));
+    else if (py::isinstance<py::slice>(val))
+      parsed.slices.push_back(
+          get_slice_range(self, {dim, py::cast<py::slice>(val)}));
+    else if (py::isinstance<Variable>(val)) {
+      if constexpr (std::is_same_v<T, Variable>)
+        throw except::DimensionError(
+            "Label-based indexing requires coordinates and is not supported "
+            "for Variable. Use a DataArray or positional indices instead.");
```
Comment (jl-wynen) [inline lib/python/bind_slice_methods.h]:
```suggestion
            "for Variable. Use positional indices instead.");
```
`self` is a Variable here, suggesting to use a `DataArray` doesn't make a lot of sense here.

File: lib/python/bind_slice_methods.h [hunk 6]
```diff
@@ -242,6 +244,62 @@ T slice_by_list(const T &obj,
...
+      else
+        parsed.slices.push_back(std::make_from_tuple<Slice>(
+            get_slice_params(self, dim, py::cast<Variable>(val))));
+    } else if (auto arr = try_cast<std::vector<scipp::index>>(val))
+      parsed.int_arrays.emplace_back(dim, std::move(*arr));
+    else if (auto i = try_cast<scipp::index>(val)) // e.g. numpy integers
```
Comment (jl-wynen) [inline lib/python/bind_slice_methods.h]: Do we need the check for `py::int_` above? Seems like this branch would handle that case.

File: src/scipp/core/data_group.py [hunk 7]
```diff
@@ -156,11 +156,19 @@ def __getitem__(self, name: Any) -> Any:
...
+        is equivalent to ``dg['x', 0]['y', 1]``.
         """
         from .bins import Bins
 
         if isinstance(name, str):
             return self._items[name]
+        if isinstance(name, dict):
+            out = DataGroup(self)
+            for dim, index in name.items():
+                out = out[dim, index]
```
Comment (jl-wynen) [inline src/scipp/core/data_group.py]: This is technically not the same as for variables and data arrays because it does not reorder array indices to the end. I think it only affects performance and makes no difference to the final result. Is that correct?

File: lib/python/bind_slice_methods.h (hunk 4 above)
Comment (SimonHeybrock) [reply lib/python/bind_slice_methods.h]: Good catch on the asymmetry, though the example doesn't quite hold: `np.str_` subclasses `str`, so `py::isinstance<py::str>` already accepted it. The case that did fail is `bytes` -- pybind11's `std::string` caster accepts those, so `x[b'abc', 0]` worked while `x[{b'abc': 0}]` raised.

Applied your helper, which removes that divergence along with the duplicate check. Added a test with an `np.str_` key anyway, since it's the more likely thing for someone to hit.

File: lib/python/bind_slice_methods.h (hunk 6 above)
Comment (SimonHeybrock) [reply lib/python/bind_slice_methods.h]: No, removed. `try_cast<scipp::index>` covers plain ints, and pybind11's integer caster rejects floats even with conversion enabled, so `{'x': 0.5}` still raises.

The one behaviour change is for ints too large for `scipp::index`: those now give the same `TypeError` as tuple-based indexing rather than escaping as a `cast_error`.

File: src/scipp/core/data_group.py (hunk 7 above)
Comment (SimonHeybrock) [reply src/scipp/core/data_group.py]: Almost -- there is one case where it changes the result rather than just performance. Slicing one dimension first can reduce a multi-dimensional coord to 1-D, which makes a label-based lookup along another dimension succeed:

```python
dg[{'y': 0, 'x': sc.scalar(1.0, unit='m')}]  # worked
dg[{'x': sc.scalar(1.0, unit='m'), 'y': 0}]  # DimensionError
```

DataArray raises for both, since it resolves every index against the unsliced object.

Since dict keys carry no meaningful order, I'd rather remove the order dependence than reorder on top of it. DataGroup now forwards the dict, restricted to the dims each item has, to the item itself, so the C++ path stays the single definition of the semantics and DataGroup inherits the array-last ordering too. `Bins` keeps the loop as it has no dict overload.

**Review by jl-wynen — APPROVED** [review]

**Comment by doronbehar** [comment]
Thanks a lot @SimonHeybrock  !
